### PR TITLE
D8CORE-2150: Adjust link widget to mirror the php link widget

### PR DESCRIPTION
**Pull request by pookmish**
# READY FOR REVIEW

# Summary
- Mirrored the functions to their PHP counterparts with a little extra in there.

# Need Review By (Date)
- 10/2

# Urgency
- medium

# Steps to Test
1. Checkout this branch & clear browser caches
1. Add a paragraph that has a link field
1. paste a url
1. validate the url you pasted is unchanged, unless it a local url in which it converts it to relative.
1. try some other urls and some autocomplete suggestions to verify it all works and saves correctly.

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

File: js/src/Components/Widgets/LinkWidget.js [hunk 1]
```diff
@@ -58,77 +59,196 @@ export const LinkWidget = ({fieldId, defaultValue, onFieldChange, settings}) =>
...
+   * @see \Drupal\link\Plugin\Field\FieldWidget\LinkWidget::getUriAsDisplayableString()
+   *
+   * @param uri
+   * @returns {*}
+   */
   const getUriAsDisplayableString = (uri) => {
-    const parsedUri = parseUrl(uri);
-    var displayString = uri;
+    const scheme = parseUrl(uri, 'PHP_URL_SCHEME');
+    console.log(urlSuggestions);
```
Comment (sherakama) [inline js/src/Components/Widgets/LinkWidget.js]: Remove console.log

File: js/src/Components/Widgets/LinkWidget.js [hunk 2]
```diff
@@ -58,77 +59,196 @@ export const LinkWidget = ({fieldId, defaultValue, onFieldChange, settings}) =>
...
-        userString = '/' + userString.substr(7);
+    // Detect a schemeless string, map to 'internal:' URI.
+    else if (string.length > 0 && parseUrl(string, 'PHP_URL_SCHEME') === undefined) {
+      // @todo '<front>' is valid input for BC reasons, may be removed by
+      //   https://www.drupal.org/node/2421941
+      // - '<front>' -> '/'
+      // - '<front>#foo' -> '/#foo'
+      if (string.indexOf('<front>') === 0) {
+        string = '/' + string.substr(7);
       }
```
Comment (sherakama) [inline js/src/Components/Widgets/LinkWidget.js:178]: You're likely going to have to handle the `<none>` syntax as well.

The app crashes when I put in a `<none>`

```
TypeError: n.ltrim is not a function
```

Steps to reproduce:
1. From a fresh build of 1.x cardinalsites, Check out this branch
2. Create new basic page
3. Create a new card
4. Enter `<none>` for the link option and unfocus off the field.
5. Crash pants.

**Review by sherakama — CHANGES_REQUESTED** [review]
Fixs

File: js/src/Components/Widgets/LinkWidget.js (hunk 2 above)
Comment (pookmish) [reply js/src/Components/Widgets/LinkWidget.js:178]: Thanks for catching that. the Crash Pants is a no good. FWIW, the data is saved correctly without breaking, but the card twig renders out an `<a href="" class="su-link su-button">Some Text</a>` without any href. But that's a bug for another area.

**Comment by pookmish** [comment]
Fixes have been implemented. Should be good now @sherakama

**Review by sherakama — APPROVED** [review]
GTG

**Pull request merged**